New docs website / Page "sidebar" styling (and other polishing)

didoo opened the pull request:
### :pushpin: Summary

This PR styles the "sidebar" elements according to the design specs. It's built on top of  #807, #840 and #844 (it's been closed in favour of a single PR, this one)

### :hammer_and_wrench: Detailed description

In #844 I have:
- added event handling logic between “burger” menu in the header, the main “application” controller, and the “sidebar”
- refactored breakpoints mixins to simplify their definitions and make them more easy to use
- added `z-index` scale as CSS tokens
- added links to “top level” routes in the sidebar navigation (only for small viewports)
- updated page “header” to use the updated breakpoint mixins
- added responsive behaviour to the “sidebar” (differentiated between homepage an other pages)
  - show/hide via "burger" menu on small viewports
  - fixed on medium/large viewports
  - for the homepage, always show/hide via "burger" menu
  - allowed the sidebar to scroll independently from the content

In this PR I have: 
- done a first pass at styling the "sidebar" elements following the current design specs
  - to be discussed/reviewed with @heatherlarsen because I am not sure what is the "final" design
- did a quick change to make the page "side**c**ar" sticky when the page is scrolled

After a review session with @heatherlarsen I have then:
- updated breakpoint for when we show/hide the sidebar in “mobile” mode
  - changed from `1000px` to `880px`, value determined testing in a browser
- made the variable `--doc-page-sidebar-width` responsive
  - at small/medium viewport the sidebar is slightly smaller to optimize the available space
- changed `max-height` value for the homepage illustration on large viewports
- removed GitHub icon from top-route link in sidebar
- refactored sidebar code to split home vs other pages + implemented home-specific animated version with gradient
- implemented collapsible “folder” in sidebar using `<details>`

👉 👉 👉 **Previews**:
- homepage: https://hds-website-git-new-docs-website-sidebar-styling-hashicorp.vercel.app/
- details page: https://hds-website-git-new-docs-website-sidebar-styling-hashicorp.vercel.app/components/alert/

### 🚧 TODOs
- ~~add a reset for the `showSidebarOnSmallViewport` on route change~~
  - tracked here: https://hashicorp.atlassian.net/browse/HDS-1258
- [ ] understand if we really need to handle overflowing of the content, or is an edge case and we can live without it for now

### :link: External links

Jira tickets::
- https://hashicorp.atlassian.net/browse/HDS-934 (main ticket)
- https://hashicorp.atlassian.net/browse/HDS-1253 (behaviour)
- https://hashicorp.atlassian.net/browse/HDS-1220 (styling)
Figma file: https://www.figma.com/file/Ky0qWjvHZR3je1lCBlzNB1/HDS-website?node-id=188%3A18330

***

### 👀 Reviewer's checklist:

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

Dhaulagiri commented on website/app/controllers/application.js:
```diff
@@ -0,0 +1,16 @@
+
+  @action
+  onToggleBurgerMenu(isOpen) {
+    console.log('onToggleBurgerMenu', isOpen);
```
```suggestion

```

Dhaulagiri commented:
question that i don't see answered in the PR description. on a page like [patterns](https://hds-website-hhgnwqsuk-hashicorp.vercel.app/patterns/) where there is no content in the sidebar, it's now visible but in current production it is not. Is that intentional?

Dhaulagiri commented:
On [components](https://hds-website-hhgnwqsuk-hashicorp.vercel.app/components/) I got into a state where I couldn't scroll the content of the main body but the sidebar was still scrollable. I could only reproduce it with the following steps:

1. Visit [components](https://hds-website-hhgnwqsuk-hashicorp.vercel.app/components/) 
2. Collapse window so sidebar is hidden
3. Open sidebar via hamburger menu
4. Resize page back to full view
5. Sidebar scrolls, main body doesn't

Re-narrowing the screen and closing the sidebar fixes the issue, so maybe there's some class or logic that's happening when the sidebar is open to prevent main body scrolling that isn't getting cleaned up this case?

didoo commented:
> On [components](https://hds-website-hhgnwqsuk-hashicorp.vercel.app/components/) I got into a state where I couldn't scroll the content of the main body but the sidebar was still scrollable. I could only reproduce it with the following steps:
> 
> 1. Visit [components](https://hds-website-hhgnwqsuk-hashicorp.vercel.app/components/)
> 2. Collapse window so sidebar is hidden
> 3. Open sidebar via hamburger menu
> 4. Resize page back to full view
> 5. Sidebar scrolls, main body doesn't
> 
> Re-narrowing the screen and closing the sidebar fixes the issue, so maybe there's some class or logic that's happening when the sidebar is open to prevent main body scrolling that isn't getting cleaned up this case?

Yes, we noticed this bug today with @alex-ju. Do you mind creating a bug and assign it to him (we discussed a possible way to fix it). Thanks

didoo commented:
> question that i don't see answered in the PR description. on a page like [patterns](https://hds-website-hhgnwqsuk-hashicorp.vercel.app/patterns/) where there is no content in the sidebar, it's now visible but in current production it is not. Is that intentional?

Something I have to think about. It's not simple to solve. Also, in theory there should be content there. So... it's temporary. I'll think about it more, maybe I can come up with something.

Dhaulagiri commented:
Opened https://hashicorp.atlassian.net/browse/HDS-1301 for the scrolling issue. I think it's mostly a nice to have given it can be fixed by reloading the page.

> Also, in theory there should be content there. So... it's temporary.

Makes sense. I think for launch we can either fix by updating the content or just live with it for now.

MelSumner reviewed (approved):
Let's ship it if we can, we can iterate.  👍

didoo commented:
> 1. When the menu is open, I can't scroll through the page's main content. This is happening on all pages.
Known issue, tracked in https://hashicorp.atlassian.net/browse/HDS-1301

> 2. On the components overview page, in the sidebar, the "Form" and "Links" categories should be closed by default.

OK, changed.

> 3. On the detail pages, around 879px, the sidebar collapses, but the global navigation doesn't change until 767px. This results in a doubling up of the "main category" links. Main category links shouldn't appear in the sidebar until 767px.

Yes, my bad, wrong media query. Fixed.

> 4. [nice to have] I didn't see the transition on the homepage when the menu opens. It just appears.

Yes, I have disabled it, but I'll re-enable and track things in a follow up ticket: https://hashicorp.atlassian.net/browse/HDS-1306

> 5. Not sure if this is the correct place to ask this, but is there a way to hide "On this page" in the sidecar if there are no links available?

We'll review how to handle this case when we have more realistic content (so, in January)
